06_gpu_and_ml/text-to-video/mochi.py

```diff
@@ -28,338 +27,147 @@
+outputs = modal.Volume.from_name(VOLUME_NAME, create_if_missing=True)
 OUTPUTS_PATH = "/outputs"  # remote path for saving video outputs
 
-# We download the model using the `hf-transfer`
-# library from Hugging Face and additionally download
-# the text encoder (Google's T5 XXL) using `transformers`.
-
-# This can takes five to thirty minutes, depending on traffic
-# and network speed.
-
-# If you want to launch the download first,
-# before running the rest of the code,
-# use the following command from the folder containing this file:
-
-# ```bash
-# modal run --detach mochi::download_model
-# ```
-
-# The `--detach` flag ensures the download will continue
-# even if you close your terminal or shut down your computer
-# while it's running.
-
-download_image = (
-    modal.Image.debian_slim(python_version="3.11")
-    .pip_install(
-        "huggingface_hub",
-        "hf-transfer",
-        "torch",
-        "transformers",
-        "sentencepiece",
-    )
-    .env(
-        {
-            "HF_HUB_ENABLE_HF_TRANSFER": "1",
-            "HF_HOME": str(MODEL_CACHE / "huggingface"),
-        }
-    )
-)
-
-image = image.env(  # so we look for the model in the right place
-    {"HF_HOME": str(MODEL_CACHE / "huggingface")}
-)
-
-
-@app.function(
-    volumes={MODEL_CACHE: model},
-    timeout=2 * HOURS,
-    image=download_image,
-)
-def download_model(
-    model_revision: str = "8e9673c5349979457e515fddd38911df6b4ca07f",
-):
-    from huggingface_hub import snapshot_download
-    from transformers import T5EncoderModel, T5Tokenizer
+MINUTES = 60
+HOURS = 60 * MINUTES
 
-    model.reload()
-    print("🍡 downloading Mochi model")
+# ## Setting up our Mochi class
+# We'll use the `modal.cls` decorator to define a [Modal Class](https://modal.com/docs/guide/lifecycle-functions)
+# which we use to control the lifecycle of our cloud container.
+# 
+# We configure it to use our image, the distributed volume, and a single H100 GPU.
 
-    snapshot_download(
-        repo_id="genmo/mochi-1-preview",
-        local_dir=MODEL_CACHE / "mochi-1-preview",
-        revision=model_revision,
+with image.imports():
+    import torch
+    from diffusers import MochiPipeline
+    from diffusers.utils import export_to_video
+    
+@app.cls(
+    image=image,
+    volumes={
+        OUTPUTS_PATH: outputs,  # videos will be saved to a distributed volume
+    },
+    gpu=modal.gpu.H100(count=1),
+    timeout= 1 * HOURS,
     )
-    print("🍡 model downloaded")
-
-    print("🍡 downloading text encoder")
-
-    T5Tokenizer.from_pretrained("google/t5-v1_1-xxl", legacy=False)
-    T5EncoderModel.from_pretrained("google/t5-v1_1-xxl")
-
-    model.commit()
-    print("🍡 text encoder downloaded")
+class Mochi:
+    # We can stack the `modal.build` and `modal.enter` decorators
+    # because we use the diffusers library, which caches the model to disk on its first run.
+    # This builds our model weights into the container image, ready for future use.
```

Comment:
We actually want to move away from this pattern, c.f. the updated guide [here](https://modal.com/docs/guide/model-weights).

In particular, we want to stop stacking `build` and `enter` most of the time, because it is confusing to users and adds extra weight to the build step (e.g. the `self.pipe` method calls on L92-3).

Reply:
We'd prefer to add the weights to a `Volume` instead. We don't have a great pattern for that with a pure `modal deploy`, but in this case we can just add it to the `local_entrypoint`. Use a separate Function to do the download. You can re-use the code in the Mochi `cls` if you want to still take advantage of HF's auto-loading/caching.